{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Quantifying the Abundance of Different Microbes in Your Metagenome using OneCodex:\n",
    "\n",
    "#### INPUT to terminal:\n",
    "\n",
    "onecodex login\n",
    "    \n",
    "#### OUTPUT from terminal:\n",
    "\n",
    "Uploading: 160523Alm_D16-4705_2_sequence.fastq.gz finished.\n",
    "\n",
    "Uploading: All complete.\n",
    "\n",
    "#### To upload two files from bioe131 onto onecodex, we used the following input into terminal:\n",
    "\n",
    "screen\n",
    "\n",
    "onecodex upload 160523Alm_D16-4705_1_sequence.fastq.gz 160523Alm_D16-4705_2_sequence.fastq.gz\n",
    "\n",
    "**Description:** this code uploaded two sequences onto onecodex and allowed the samples to be indexed, searched, and compared. Running it in a screen allowed the program to keep running even if the machine went offline.\n",
    "\n",
    "- After the sequences were uploaded onto onecodex, we went to the onecodex website to look at our results.\n",
    "\n",
    "#### Results for 160523Alm_D16-4705_1_sequence.fastq.gz are as follows\n",
    "\n",
    "| Species | Amount in Soil | Description|\n",
    "|---------|----------|------------|\n",
    "|Cutibacterium acnes|9.84%| linked to skin acne|\n",
    "|Actinomyces odontolyticus|4.66%| bacteria causing highly infectious disease|\n",
    "|Cladosporium sphaerospermum|4.33%| found in decaying leaves and branches of citrus|\n",
    "|Armatimonadetes bacterium 13_1_40CM_64_14|3.64%| found in aquatic plants|\n",
    "|Pseudomonas aeruginosa|2.75%| found in many different environments such as soil, water, and plant and animal tissue|\n",
    "|Streptomyces californicus|2.58%|originates in moist surfaces|\n",
    "|Micrococcus luteus|2.35%| found in soil, dust, water and air, and as part of the normal flora of the mammalian skin|\n",
    "|Escherichia coli|2.02%| commonly found in the lower intestine of warm-blooded organisms|\n",
    "\n",
    "#### *The following species were also found in the sample but at less than 2% concentration.\n",
    "| Species | Amount in soil|\n",
    "|-------|-------|\n",
    "|Nocardioides sp. REDSEA-S25_B9|1.83%|\n",
    "|Moraxella osloensis|1.79%|\n",
    "\n",
    "#### Results for 160523Alm_D16-4705_2_sequence.fastq.gz are as follows: \n",
    "\n",
    "| Species | Amount in soil| Description|\n",
    "|-----------|--------------|------------|\n",
    "|Cutibacterium acnes | 9.91% | linked to skin acne|\n",
    "|Cladosporium sphaerospermum | 4.33% |found in decaying leaves and branches of citrus|\n",
    "|Armatimonadetes bacterium |3.65%|found in aquatic plants|\n",
    "|Pseudomonas aeruginosa | 2.60% | found in soil, water, and plant and animal tissue| \n",
    "|Escherichia coli | 2.06%| commonly found in the lower intestine of warm-blooded organisms|\n",
    "\n",
    "#### *The following species were also found in the sample but at a less than 2% concentration.\n",
    "\n",
    "| Species | Amount in soil|\n",
    "|-----------|----------|\n",
    "|Williamsia sp. D3 |1.83%|\n",
    "|Moraxella osloensis|1.80%|\n",
    "|Williamsia muralis|1.59%|\n",
    "|Micrococcus luteus|1.52%|\n",
    "|Micrococcaceae bacterium JKS001869|1.44%|\n",
    "|Acidobacteria bacterium 13_1_20CM_2_60_10|1.43%|\n",
    "|Sphingomonas ginsenosidimutans|1.41%|\n",
    "|Asanoa ferruginea|1.36%|\n",
    "|Methylobacterium sp. CG08_land_8_20_14_0_20_71_15|1.11%|\n",
    "|Enhydrobacter sp. H5|1.09%|\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Assemble your Metagenomeinto Contigs Using the Metaspadespart of SPAdes.\n",
    "\n",
    "**INPUT to terminal:**\n",
    "    \n",
    "    spades.py --meta -t 10 -m 64 -1 /data/metagenomes/160523Alm_D16-4705_1_sequence.fastq.gz  -2 /data/metagenomes/160523Alm_D16-4705_2_sequence.fastq.gz -o /home/be131-19/FinalProject metaspades_assembly\n",
    "    \n",
    "**OUTPUT from terminal:**\n",
    "\n",
    "**Description:** The purpose of this code is to use the sequences provided and assemble the metagenomic data in order to analyze many complex bacterial populations.\n",
    "\n",
    "**Issues:** Our code did not run to completion. A potential explanation for this error occuring is due to the abundance of students occupying the cluster and thus taking up the memory. This then does not allow metaspades to run. Although we attempted to run this several times, it did not function do to this issues Going forward, we will continue to attempt to fix this issue by asking for a different data set. Hello\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
